apisix/plugins/ldap-auth.lua

```diff
@@ -0,0 +1,158 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local core = require("apisix.core")
+local ngx = ngx
+local ngx_re = require("ngx.re")
+local ipairs = ipairs
+local consumer_mod = require("apisix.consumer")
+local lualdap = require "lualdap"
+
+local lrucache = core.lrucache.new({
+    ttl = 300, count = 512
+})
+local consumers_lrucache = core.lrucache.new({
+    type = "plugin",
+})
+
+local schema = {
+    type = "object",
+    title = "work with route or service object",
+    properties = {
+        base_dn = { type = "string" },
+        ldap_uri = { type = "string" },
+        use_tls = { type = "boolean" },
+        uid = { type = "string" }
+    },
+    required = {"base_dn","ldap_uri"},
+    additionalProperties = false,
+}
+
+local consumer_schema = {
+    type = "object",
+    title = "work with consumer object",
+    properties = {
+        user_dn = { type = "string" },
+    },
+    required = {"user_dn"},
+    additionalProperties = false,
+}
+
+local plugin_name = "ldap-auth"
+
+local _M = {
+    version = 0.1,
+    priority = 2540,
+    type = 'auth',
+    name = plugin_name,
+    schema = schema,
+    consumer_schema = consumer_schema
+}
+
+function _M.check_schema(conf, schema_type)
+    local ok, err
+    if schema_type == core.schema.TYPE_CONSUMER then
+        ok, err = core.schema.check(consumer_schema, conf)
+    else
+        ok, err = core.schema.check(schema, conf)
+    end
+
+    return ok, err
+end
+
+local create_consume_cache
+do
+    local consumer_names = {}
+
+    function create_consume_cache(consumers)
+        core.table.clear(consumer_names)
+
+        for _, consumer in ipairs(consumers.nodes) do
+            core.log.info("consumer node: ", core.json.delay_encode(consumer))
+            consumer_names[consumer.auth_conf.user_dn] = consumer
```

Comment:
What if two users use the same `user_dn`?

Reply:
A [distinguish name](https://ldapwiki.com/wiki/Distinguished%20Names) (dn) is per definition unique so user cannot share the same dn

Reply:
Got it.

Reply:
But what if the admin configured the same dn for different consumers accidentally? Is it harmless or there are some side effects?

Reply:
Yes it's harmeless, is there a way in the consumer definition to identify a unique field ?

Reply:
Got it.

> is there a way in the consumer definition to identify a unique field

We have to check it by our own. It seems that the JSON schema cannot help us to do it. Anyway, we may document it.